### PR TITLE
Hard-Coded Checkpoint Lists

### LINKED ISSUE #564: Create and parse checkpoint height/hash lists
For the checkpoint verifier, we need to create checkpoint height and hash lists for:
- [x] main net, and
- [x] test net, and
- [x] parse those lists into Zebra data structures.

Since the sapling checkpoint is mandatory, we want to compile the lists into Zebra.

Eventually, we also want to:
- [ ] re-implement the list generation bash scripts in Rust (opened #684)
- [ ] support generating incremental updates to the checkpoint lists, given the final checkpoint height (opened #685)

Open questions:
- [x] Is 256 MB spacing too high? Do we need to impose a maximum number of blocks as well?
  - We decided on a maximum 2000 block spacing
- [ ] Should we remove any checkpoints after Heartwood activation?

The block spacing range for 256 MB checkpoints is:
  - Mainnet: 400 - 50,000
  - Testnet: 10,000 - 250,000

Heartwood activated at:
  - Mainnet: 903000 (16 July 2020)
  - Testnet: 903800 (4 May 2020)

Our initial draft list of checkpoints was created before Heartwood on mainnet, but after Heartwood on testnet. The testnet checkpoints include a post-Heartwood block.

### PR DESCRIPTION
**Pull request by teor2345**
I implemented the checkpoint calculations using some bash scripts, and committed them to zebra-utils. We can re-implement them in Rust when we have time.

Here are my questions:
- [x] Is 256 MB spacing too high? Do we need to impose a maximum number of blocks as well?
  - We decided on a maximum 2000 block spacing
- [ ] Should we remove any checkpoints after Heartwood activation?

The block spacing range for 256 MB checkpoints is:
  - Mainnet: 400 - 50,000
  - Testnet: 10,000 - 250,000

Heartwood activated at:
  - Mainnet: 903000 (16 July 2020)
  - Testnet: 903800 (4 May 2020)

Our initial draft list of checkpoints was created before Heartwood on mainnet, but after Heartwood on testnet. So the testnet checkpoints include post-Heartwood blocks.

Closes #564.

### CODE REVIEW

**Comment by teor2345** [comment]
The first new commit is c561827.

(The commit will change when #663 and #670 are updated or merged.)

**Comment by hdevalence** [comment]
I think that the block spacing might be too long.  I think that setting the sync lookahead limit to twice the maximum checkpoint distance is a good way to be safe while downloading blocks (choosing 2*d rather than d guards against aliasing the checkpoint interval), and downloading 100k blocks at a time seems less than ideal.  We could try to do some plumbing to feed data from the checkpoint verifier to the syncer, but that seems much more complex compared to hardcoding additional hashes.

My guess is that a 2000 block distance limit would be reasonable to start.

It would be convenient if the bash script did the hex-reversal, even if it's somewhat slower.

**Comment by teor2345** [comment]
> It would be convenient if the bash script did the hex-reversal, even if it's somewhat slower.

The calculate-checkpoints.sh bash script does the hex reversal.

Doing the reversal on every block hash would take at least 15 hours, but it should be a lot faster after #679 is merged. For the full calculation, see:
https://github.com/ZcashFoundation/zebra/pull/670#issuecomment-659668194

**Comment by teor2345** [comment]
I rebased onto main after the PRs merged.

> My guess is that a 2000 block distance limit would be reasonable to start.

See 1df13a8f9b8a10bf79d14d5b931f783f3b0da329 for the script update, and 9774315b08a62fd00dd60b59c5a35dfb4cbc6e1c for the new lists.

I squashed the list updates together, because they didn't have any checkpoints in common. So the diffs weren't useful.

**Comment by teor2345** [comment]
Oh, and I still have this open question:
- [ ] Should we remove any checkpoints after Heartwood activation?

**Comment by hdevalence** [comment]
Great! I'd be happy to merge this whenever.  I think it would be nice if we could generate Rust source code directly but maybe this is better done later -- personally I think it would be fine to create the initial checkpoint list by copypasting the text files into a `.rs` and filling in syntax.

**Comment by teor2345** [comment]
TODO:
- [x] edit the scripts to reference LOOKAHEAD_LIMIT
- [x] parse the checkpoint lists in the tests, so I can check they actually work

Then I think we can merge.

**Comment by teor2345** [comment]
> (choosing 2*d rather than d guards against aliasing the checkpoint interval)

Can you explain this a bit more?
I'm not sure exactly what you mean.

**Comment by hdevalence** [comment]
@teor2345 I didn't think it through precisely, but I was thinking of something like the Nyquist limit, with the checkpoint frequency aliasing the lookahead frequency (? or something?) but I'm not sure this makes sense.  However if we do `2*d` rather than `d` then the syncer will attempt to download the next interval while the current interval is downloading, so that we don't pause downloads.

**Comment by teor2345** [comment]
I think it's a good idea. It should help avoid stalls, and a whole class of similar issues, such as off-by-one errors and thundering herds.

**Comment by teor2345** [comment]
I parsed the lists, and added some tests.

I also tweaked the CheckpointList and CheckpointVerifier initialisation functions, so callers can just ask for a verifier with the hard-coded list.

File: zebra-consensus/src/checkpoint/list.rs [hunk 1]
```diff
@@ -30,31 +34,72 @@ type Error = Box<dyn error::Error + Send + Sync + 'static>;
...
+    /// Assumes that the provided genesis checkpoint is correct.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut checkpoint_list: Vec<(BlockHeight, BlockHeaderHash)> = Vec::new();
+
+        for checkpoint in s.lines() {
+            let fields = checkpoint.split(' ').collect::<Vec<_>>();
+            if let [height, hash] = fields[..] {
+                checkpoint_list.push((height.parse()?, hash.parse()?));
+            } else {
+                Err(format!("Checkpoint list parse failure. Expected 2 space-separated fields on checkpoint line: '{}'", checkpoint))?;
```
Comment (yaahc) [inline zebra-consensus/src/checkpoint/list.rs]:
```suggestion
                Err(format!("Invalid checkpoint format: expected 2 space-separated fields but found {}: '{}'", fields.len(), checkpoint))?;
```

**Review by yaahc — CHANGES_REQUESTED** [review]
Looks good, I didn't carefully review the bash because I assume we're going to rewrite it in rust soon but the rust looks good. There is one error message I'd like to update but other than that this looks good to me. I like the bit where we use the first checkpoint in the list to determine the network it should be from :D

**Comment by teor2345** [comment]
> I didn't carefully review the bash because I assume we're going to rewrite it in rust soon

Does anyone carefully review bash? 😄

I think we're pretty safe here, because the CheckpointVerifier is very flexible. The exact heights aren't important, just as long as the heights and hashes match, and there is a genesis checkpoint.

> I like the bit where we use the first checkpoint in the list to determine the network it should be from :D

"Make invalid states un-representable", but for function arguments.

Daira previously suggested that we should use the actual consensus parameters as much as possible. (Rather than passing a network enum around everywhere.) So I removed or replaced most network references, and converted from the network as soon as I could. Right now, nothing in `zebra-consensus` keeps a persistent copy of the network.

**Comment by teor2345** [comment]
(I merged this because the last change was trivial, and I wanted to make new PRs based on this PR.)

**Pull request merged**